### PR TITLE
Per framework Xcode projects/dev environments, Podfiles, updated/pushable Podspecs

### PR DESCRIPTION
**Pull request by mhuusko5**
Issue (was late about it, sorry): https://github.com/firebase/firebase-ios-sdk/issues/21

EDIT: More to do with #23 / #9. macOS component superseded by https://github.com/firebase/firebase-ios-sdk/pull/38.

The main goal of this effort was to reintroduce macOS support. As part of that process though, it became clear that the mono podspec/project (or really, just tests project, no dev project at all) approach was holding things back, so I first introduced individual xcode projects/builds for each framework, separate podfiles to pull in their dependencies, and updated each's imports and podspecs along the way to make the various build methods work happily together.

The end result is..
Dev environments for each framework
Potential step towards Carthage support (Carthage needs an .xcodeproj to build)
All the frameworks can be built separately locally, separately by hosting the new podspecs (or locally/hosted with the mono FirebaseDev podspec as they could before), on iOS or macOS (except for Messaging), library or framework, Obj-C or Swift

### CODE REVIEW

**Comment by mhuusko5** [comment]
I signed it!

**Comment by paulb777** [comment]
Wow! Thanks for the substantial contribution!

We'll get a review going in the next few days.

File: Firebase/Auth/Source/AuthProviders/Phone/FIRPhoneAuthProvider.m [hunk 1]
```diff
@@ -15,18 +15,19 @@
  */
 
 #import "FIRPhoneAuthProvider.h"
+#import "FirebaseAuth_Internal.h"
 
-#import "FIRLogger.h"
+#import FirebaseCore(FIRLogger.h)
```
Comment (paulb777) [inline Firebase/Auth/Source/AuthProviders/Phone/FIRPhoneAuthProvider.m]: Why the FirebaseCore macro? I can see that it makes import sources clearer. Any references about this style?

File: Firebase/Auth/FirebaseAuth.podspec [hunk 2]
```diff
@@ -20,37 +20,60 @@ Simplify your iOS development, grow your user base, and monetize more effectivel
...
+    'Firebase/Auth/CHANGELOG.md', 'CHANGELOG.md'
+
   s.xcconfig = { 'GCC_PREPROCESSOR_DEFINITIONS' =>
     '$(inherited) ' + 'FIRAuth_VERSION=' + s.version.to_s +
     ' FIRAuth_MINOR_VERSION=' + s.version.to_s.split(".")[0] + "." + s.version.to_s.split(".")[1]
   }
+
   s.framework = 'Security'
-#  s.dependency 'FirebaseDev/Core'
+  s.dependency 'FirebaseCore', '~> 4.0.1'
```
Comment (paulb777) [inline Firebase/Auth/FirebaseAuth.podspec]: This won't be publishable as long as the release version of FirebaseCore has closed source dependencies -
 https://github.com/CocoaPods/Specs/blob/master/Specs/8/b/d/FirebaseCore/4.0.0/FirebaseCore.podspec.json

File: Firebase/Auth/FirebaseAuth.podspec [hunk 3]
```diff
@@ -20,37 +20,60 @@ Simplify your iOS development, grow your user base, and monetize more effectivel
...
+      'FIRAuthNotificationManager.[mh]',
+      'FIRAuthAPNSTokenManager.[mh]',
+      'FIRAuthAPNSTokenType.[mh]',
+      'FIRAuthAPNSToken.[mh]',
+      'AuthProviders/Phone/FIRPhoneAuthProvider.[mh]'
+    ]]
+  end
 
-  s.source_files = '**/*.[mh]'
-  s.public_header_files =
```
Comment (paulb777) [inline Firebase/Auth/FirebaseAuth.podspec]: Removing the public headers specifires  is a problem. I'm seeing all the headers published in the framework instead of only the public ones, when I build with https://github.com/mhuusko5/firebase-ios-sdk/blob/master/BuildFrameworks/build.swift

File: Firebase/Auth/Source/FirebaseAuthVersion.m [hunk 4]
```diff
@@ -20,7 +20,7 @@
 #define STR(x) STR_EXPAND(x)
 #define STR_EXPAND(x) #x
 
-const double FirebaseAuthVersionNumber = FIRAuth_MINOR_VERSION;
+const double FirebaseAuthFrameworkVersionNumber = FIRAuth_MINOR_VERSION;
```
Comment (paulb777) [inline Firebase/Auth/Source/FirebaseAuthVersion.m]: Why this rename?

File: Firebase/Core/Podfile [hunk 5]
```diff
@@ -0,0 +1,31 @@
...
+end
+
+target :FirebaseCore_macOS do
+  project 'FirebaseCore.xcodeproj'
+  platform :osx, '10.10'
+
+  shared_pods
+end
+
+post_install do |installer|
```
Comment (paulb777) [inline Firebase/Core/Podfile]: What's the purpose of the post_install?

File: Firebase/Auth/Source/Private/FirebaseAuth_Internal.h [hunk 6]
```diff
@@ -0,0 +1,23 @@
+/*
```
Comment (paulb777) [inline Firebase/Auth/Source/Private/FirebaseAuth_Internal.h]: Is there a way to share the new _Internal.h/_Private files, since they're the same? At least they should be consistently named. The others are all _Private.h.

File: Firebase/Database/Source/Login/FAuthTokenProvider.m [hunk 7]
```diff
@@ -48,7 +48,7 @@ @protocol FIRFirebaseAuthLike <NSObject>
  * This is a hack that copies the definitions of Firebear error codes. If the error codes change in the original code, this
  * will break at runtime due to undefined behavior!
  */
-typedef NS_ENUM(NSUInteger, FIRErrorCode) {
+typedef NS_ENUM(NSInteger, FIRErrorCode) {
```
Comment (paulb777) [inline Firebase/Database/Source/Login/FAuthTokenProvider.m]: Why this change?

**Comment by mhuusko5** [comment]
@paulb777 thanks for the speedy feedback! Timezone dictates that I won't get

**Comment by mhuusko5** [comment]
@paulb777 thanks for the speedy feedback! Timezone dictates that I won't get to the details until tomorrow, but I will preface them now by saying all the changes in this PR were from starting with these new isolated builds, targeting iOS/macOS, then tinkering the warnings/errors away. So some of the changes/hacks are weird, but so is the result/goal of FirebaseDev being buildable as a single pod/pile of source while also the having the individual frameworks being buildable as separate pods (with imports of internal headers from Core). Maybe there's another way, maybe it's not even a desired goal (I am interested to hear the overall planned deployment strategy); if so, it was an interesting exercise (/way to familiarize)!

I'll add an example project tomorrow that can serve to demonstrate macOS, demonstrate the interop of the updated podspecs, and serve as some context for discussion of issues I was seeing that lead to the various changes. 

As for the unit tests/example mono project and macOS, I'd need to dig a bit to find out what that would take. To come!

**Comment by paulb777** [comment]
@mhuusko5 Here's some background on the pod structure:

First of all, it's very much still a work in progress and one of the reasons we haven't yet open sourced the build and release process, which is something like the [build.swift script](https://github.com/mhuusko5/firebase-ios-sdk/blob/master/BuildFrameworks/build.swift). We'd like to solve two problems for which we haven't found ideal solutions with CocoaPods:

1.  The open source pods have product dependencies on closed source libraries/pods that we're not currently able to open source.

1.  We'd like to continue to build the pods as static library frameworks to avoid the cold start impact of dynamic frameworks. More details at https://github.com/CocoaPods/CocoaPods/issues/6651

The reason for the FirebaseDev monopod is so that when a Podfile specifies !use_frameworks, the Firebase pods get consolidated via the chosen subspecs into a single dynamic library framework instead of one dynamic framework for each component with the associated several hundred millisecond impact on cold start time.

**Comment by mhuusko5** [comment]
@paulb777 I'm actually probably not going to get back to this PR until a bit later in the week unfortunately. In the meantime, this deployment strategy topic is interesting/a bit of conundrum, and I'd like to make sure I've got the full picture:

On 1) just to confirm, there's not anything inherently problematic CocoaPods-wise about having Core/Database/Auth/etc. open source, while Analytics remains closed source, right? The open source ones could be deployed as source based pods (or mono pod with subspecs), or vendored frameworks based pods, and have a dependency on the definitely vendored framework based Analytics pod, as they currently do. In fact, if I understand correctly, they wouldn't even necessarily need a dependency on the Analytics pod (user's could just pull it in alongside if they desired), because it's a weak/optional dependency, but it's just optimal for complete product exposure for it to be included by default..? That's a long way of saying I assume the source being open doesn't have to change the deployment strategy (assuming your previous build/deployment strategy didn't have intricacies/assumptions between the source that is now open, and that which is still closed)? 

2) Static vs. dynamic (or number of dynamics) seems to be the hot ticket. It seems though that there may not be many (good) options..
    1. Analytics static. The other frameworks built statically too, distributed as vendored framework pods separately. I imagine this was/is the 3.x setup? Not a bad option, just isn't great for those wrapping Firebase. 
    2. Analytics static. Others mono source/dynamic framework with a dependency on it. If user's want to use the static  Analytics, they have to get creative (https://github.com/CocoaPods/CocoaPods/issues/2926). Not a great option.
    3. Analytics dynamic. Others mono source/dynamic or separate source/dynamics. Simple/flexible, but with launch cost – that said, is it really 600ms per framework or does it require specific conditions (I work on an integration heavy use_frameworks! app with 33 direct/transitive pods, and our launch time is ~3 seconds; not great, but not 20 seconds either)?

**Comment by paulb777** [comment]
@mhuusko5 Thanks for the update on your timing and let us know if you'd like more clarity on anything else.

1) The pods could be deployed as source only if (A) we removed the dependency on Analytics (since CocoaPods does not support source pods depending on static library pods) or (B) we built a dynamic library version of Analytics or (C) we removed the dependency. 

We deploy them as vendored framework pods now. The downside is the disconnect between the source and the product build.

Right, the source being open doesn't have to change the deployment strategy. However, we'd like to make sure that we're fully understanding, mitigating any impact, and potentially reconsidering any deployment strategy constraints that are hindering open source quality.

2i. Yes, all static vendored_frameworks is both the 3.x and the 4.x product distribution scheme.

ii. Right, CocoaPods does not allow source libraries to depend on static pods. We would need to determine if "def installer.verify_no_static_framework_transitive_dependencies" or something else is a solution.

iii. Our measurements might actually be comparable. We were seeing 500-700 msec on an iPhone 6. Smaller libraries were faster than bigger ones, but still more than a half second. Newer devices are substantially faster. However, Firebase won't be viable for many potential apps if it doesn't support old phones reasonably well.

**Comment by mhuusko5** [comment]
@paulb777 The distribution topic much of the changes in this PR touch is larger than I initially understood and I don't want it to hold up macOS support, so I have started https://github.com/firebase/firebase-ios-sdk/pull/38 to cover the macOS compatibility changes in isolation, as well as getting the example/test suite and static build targeting macOS.

I'm going to close this PR for now, and maybe come back to the branch to salvage some of the organisational changes, depending on what direction the distribution goes.
